### ISSUE TITLE
[REVIEW]: Integrated hydrologic model development and postprocessing for GSFLOW using pyGSFLOW

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3852]
**Submitting author:** @jlarsen-usgs (<a href="http://orcid.org/0000-0002-1218-800X">Joshua Larsen</a>)
**Repository:** https://github.com/pygsflow/pygsflow.git
**Branch with paper.md** (empty if default branch): 
**Version:** 1.0.2
**Editor:** @crvernon
**Reviewers:** @thurber, @mdbartos, @mdbartos
**Archive:** 10.5281/zenodo.6468426

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92/status.svg)](https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92"><img src="https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92/status.svg)](https://joss.theoj.org/papers/d3989e65aedcc241930001cd32ba3a92)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/pygsflow/pygsflow.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@thurber & @mdbartos, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @crvernon know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @thurber
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/pygsflow/pygsflow.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/pygsflow/pygsflow.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jlarsen-usgs) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @mdbartos
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/pygsflow/pygsflow.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/pygsflow/pygsflow.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jlarsen-usgs) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by crvernon** [comment]
:mega: Mid-week rally!  Looks like @thurber has started providing feedback to @jlarsen-usgs.  Let's keep things rolling towards a successful review!

:wave: @mdbartos - please update me to your progress thus far and don't hesitate to reach out if you have any questions.

👏 Keep up the good work!

**Comment by jlarsen-usgs** [comment]
review from @thurber is linked to 

pygsflow/pygsflow#12

**Comment by crvernon** [comment]
:mega: Mid-week rally!  Looks like issues are cycling with @thurber and @jlarsen-usgs.  Let's keep things rolling towards a successful review!

:wave: @mdbartos - Please let me know if you are unable to participate in this review at this time.  If you are still willing, could you reply here with an expected time of completion?  Thanks!

**Comment by mdbartos** [comment]
@crvernon Can I get permissions to edit my checklist? Thx.

**Comment by mdbartos** [comment]
Could I get another invite? I still don't have permissions to edit the checklist.

**Comment by crvernon** [comment]
@whedon add @mdbartos as reviewer

**Comment by whedon** [comment]
OK, @mdbartos is now a reviewer

**Comment by crvernon** [comment]
Ok, @mdbartos you should have just received another invite to review this submission.  After you accept, you will have access to the checklist.  Let me know if you have any further questions.

**Comment by crvernon** [comment]
@mdbartos were you able to confirm and then access the checklist?

**Comment by mdbartos** [comment]
Yes, it is working now. Thank you.

**Comment by thurber** [comment]
hi @crvernon & @jlarsen-usgs, I've completed my review checklist and am satisfied with the resolutions of (pygsflow/pygsflow#12)[https://github.com/pygsflow/pygsflow/issues/12]. There's one unchecked box remaining, but i don't believe it to be critical path. Nice work!

**Comment by jlarsen-usgs** [comment]
Thanks @thurber! 

I have been waiting to update the text fully until all of the review comments are in (journal reviews and internal USGS). The unchecked box in pygsflow#12 will be addressed in the final version.

**Comment by crvernon** [comment]
Thanks @thurber !

@mdbartos do you have an expected date that you could have your review completed by?  Thanks!

**Comment by mdbartos** [comment]
I should be able to get through the checklist items this weekend, although there are some issues with the installation depending on which branch/python version I am using.

To be clear, should I be reviewing the `develop` branch of the current github repo?

**Comment by mdbartos** [comment]
Greetings, I've completed as much of my review as I can as of right now.

Currently I cannot install the software from the `develop` branch.

I can install from pypi and from the `master` branch. However, when I install this way, the unit tests are currently failing.

These issues need to be resolved before I can move further.

Thanks,
MDB

**Comment by mdbartos** [comment]
Welp, I was trying to install on a mac. So now it makes sense why the unit tests were failing on master.

Which branch does the JOSS release correspond to? Assuming it's `develop`, would it not make more sense to wait until an official release of flopy 3.3.5 is uploaded to pypi before publishing? Ideally, the installation procedure should be seamless for the end user. Especially if you are formally publishing a piece of software, you don't want users to be manually installing dependencies that are still in development, right?

**Comment by jlarsen-usgs** [comment]
@mdbartos 

Instead of waiting until the next release of FloPy (which is most likely going to be sometime in late February), I went ahead and made some changes to `develop` that provide support for both FloPy 3.3.4 (current release) and FloPy's 3.3.5 develop version.

**Comment by crvernon** [comment]
:mega: Mid-week rally!  Looks like we are getting close to completing the review!  @mdbartos and @jlarsen-usgs please post an update here on the status of what is remaining.  Thanks!

**Comment by mdbartos** [comment]
Greetings,

Got behind due to end-of-the semester madness, but I'm planning to install and run unit tests on my linux box and windows laptop today. If all goes well that should take care of the major outstanding issues.

MDB

**Comment by mdbartos** [comment]
Greetings,

I think I've finished the major part of my review. My only comments are the following:

- See the two minor comments on installation in my pygsflow repo issue.
- Include cleaned-up and simplified examples in the documentation pages (not just jupyter notebooks).

Aside from that, I think this is good to go.

Cheers,
MDB

**Comment by crvernon** [comment]
✅ Thanks @mdbartos !

@jlarsen-usgs once you have addressed the reviewer comments let me know and I will confirm that all is in order and make suggestions for any edits that I think are necessary to satisfy journal requirements.

**Comment by jlarsen-usgs** [comment]
Thanks for your review @mdbartos 

I'll be working through the issues that you mentioned in the coming week or so!

**Comment by crvernon** [comment]
:mega: Mid-week rally!

It would be good to get this one wrapped up soon @jlarsen-usgs so we can free up our reviewers.  How are things looking on your side of things for finishing up the remaining issues?  It looks like @mdbartos has two boxes left to check off and @thurber has approved all.

Thanks!

**Comment by jlarsen-usgs** [comment]
Hi @crvernon 

I've advanced it to the last step in the USGS Bureau Approval process that goes in conjunction with the journal review. Once I get the approval there I'll be able to check the remaining issues and have the work on my end wrapped up. Thanks for checking in, the final step on my end shouldn't take very long (< 1 week).

**Comment by crvernon** [comment]
Sounds great @jlarsen-usgs !  Let me know when you are clear and I'll make my final pass and recommendation.

**Comment by crvernon** [comment]
@jlarsen-usgs - have you gotten the approval  and remaining issues straightened out yet?  I would like to move this review forward.

Thanks!

**Comment by jlarsen-usgs** [comment]
@crvernon - Thanks for checking in, I've hit a minor snag and am going through a second USGS software security review. Once this is done, I should be able to more forward. I sincerely appologise for the delays on my end.

**Comment by mdbartos** [comment]
You can check off my review. Everything looks good to me.

**Comment by jlarsen-usgs** [comment]
@crvernon 
We're getting close to having the USGS review process complete and I'm currently expecting this to be able to move forward sometime next week. I'll let you know as soon as things wrap up on our end so we can move forward here. Thanks again for your patience!

**Comment by jlarsen-usgs** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03852/joss.03852/10.21105.joss.03852.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03852/joss.03852/10.21105.joss.03852.pdf) :page_facing_up: :point_left:

**Comment by jlarsen-usgs** [comment]
@crvernon 
We've gotten the final approval from USGS to go ahead and publish, so we can finally start wrapping this up.

**Comment by crvernon** [comment]
@whedon generate pdf

**Comment by crvernon** [comment]
@editorialbot check references

**Comment by crvernon** [comment]
@jlarsen-usgs The following should be addressed in the paper:

- [ ] Line 8.  Capitalize Python.
- [ ] Line 12.  Need citations backing the claims made in the first and second sentences.
- [ ] Line 19.  "Among these IHMs is GSFLOW..." should be "Among these IHMs is the GSFLOW..."
- [ ] Line 23.  "...larger datasets, from" should be "...larger datasets from"
- [ ] Line 32. "GSFLOW model development previously has been a..." should be "GSFLOW model development has previously been..."
- [ ] Line 38.  "integrated hydrologic models" should be "IHMs" - check this throughout the paper
- [ ] Line 51-52.  Use IHMs instead of exploded name
- [ ] Line 131.  should be "United States Geological Survey (USGS)" since it is the first use.

❗ The current license (CC0) is not allowed for use in JOSS as it is not in the list of OSI-approved licenses that we utilize (see https://opensource.org/licenses/alphabetical).  The following is some discussion on the current status of CC0 license use in the open source community if your are interested:  https://opensource.org/faq#cc-zero and https://opensource.org/faq#public-domain.  Please choose a new license to proceed.

Thanks and let me know if you have any questions!

**Comment by crvernon** [comment]
@jlarsen-usgs ... BTW, here is the USGS guidance on licensing that may help you:  https://www.usgs.gov/software-management/distribution-usgs-code 

This indicates that you should be able to use an alternative option.  Feel free to reach out if you have questions.  Thanks!

**Comment by jlarsen-usgs** [comment]
@crvernon 

I looked over that guidence earlier today and it looks like "The Unlicense" is the acceptable option that's included in the OSI approved licenses that you sent. I sent out an email earlier today to check with our policy folks to make sure there isn't any guidence that superceedes this. I'll be updating the paper based on your review comments sometime today. 

Thank you for help and your review!

**Comment by jlarsen-usgs** [comment]
@crvernon 

I've updated both the license and the paper based on your review comments. I've also updated the first paragraph of the introduction for better clarity based on your review comment about "Line 12".

**Comment by crvernon** [comment]
@editorialbot generate pdf

**Comment by crvernon** [comment]
👏 Nice work on the revisions @jlarsen-usgs 

Please check on the above missing and invalid DOIs ⬆️ and see if we can get those fixed up.  Then I will be ready to move to the next step in this process.  Also, you may want to check the spelling of "hydraulicly" on Line 21; may need two l's.

**Comment by jlarsen-usgs** [comment]
@editorialbot generate pdf

**Comment by jlarsen-usgs** [comment]
@editorialbot check references

**Comment by jlarsen-usgs** [comment]
@crvernon 

I updated "hydraulicly to hydraulically" in the paper. Thanks for catching this!

I checked the DOI. In the PDF that's generated the DOI is listed as `10.1061/(ASCE)WR.1943-5452.0000034`, which is the correct (valid) DOI. The `check references` method seems either to not be recognising or ignoring the parenthesis around (ASCE) which is causing it to report falsely that the DOI is invalid. See reference "Chiu, Y., Nishikawa, T., Yeh, W.G. ..." in the PDF of the paper for confirmation.

**Comment by jlarsen-usgs** [comment]
@crvernon 

Checking in to see if there are any other updates that need to be done before moving forward.

**Comment by jlarsen-usgs** [comment]
@editorialbot check references

**Comment by jlarsen-usgs** [comment]
@editorialbot check references

**Comment by jlarsen-usgs** [comment]
@crvernon

Reference check passes now. All DOIs are present and recognized.

**Comment by crvernon** [comment]
@jlarsen-usgs apologies for the delay. I had a family emergency that I had to tend to. I'll finalize this early next week.

**Comment by jlarsen-usgs** [comment]
@crvernon no worries, the publication can wait until you've got the time. Just tag me in a comment when you're ready to start moving this forward.

**Comment by crvernon** [comment]
@editorialbot generate pdf

**Comment by crvernon** [comment]
@jlarsen-usgs thanks for putting together a really nice software product!  Thanks to @thurber  and @mdbartos for a constructive and timely review!

I am recommending that your submission be accepted.  An EIC will review this shortly and confirm final publication if all goes well.

**Comment by jlarsen-usgs** [comment]
@crvernon 

I've put together the github release and created a Zenodo archive for this publication. The doi is [10.5281/zenodo.6468138](https://doi.org/10.5281/zenodo.6468138)

**Comment by crvernon** [comment]
@editorialbot set 1.0.2 as version

**Comment by crvernon** [comment]
@editorialbot set 10.5281/zenodo.6468138 as archive

**Comment by crvernon** [comment]
@editorialbot generate pdf

**Comment by crvernon** [comment]
@editorialbot check references

**Comment by crvernon** [comment]
@editorialbot recommend-accept

**Comment by jlarsen-usgs** [comment]
@crvernon 

I just read through the proof and found a typo that should be corrected before final acceptance and publication.  "gsflow_coef" should be "gwflow_coef". I'll correct this right now.

**Comment by jlarsen-usgs** [comment]
@crvernon 

Fixed the typos on my end.

**Comment by crvernon** [comment]
@jlarsen-usgs - no problem, please create a new release with the fix and post the new version and DOI here.  Then I will assign them accordingly.

**Comment by jlarsen-usgs** [comment]
@crvernon - created a new release in github and posted the new version to zenodo [10.5281/zenodo.6468426](https://doi.org/10.5281/zenodo.6468426)

**Comment by crvernon** [comment]
@editorialbot set 10.5281/zenodo.6468426 as archive

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@thurber, @mdbartos, @mdbartos – many thanks for your reviews here and to @crvernon for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@jlarsen-usgs – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by jlarsen-usgs** [comment]
Thank you for your helpful reviews @thurber, @mdbartos, and @crvernon!

**Comment by jlarsen-usgs** [comment]
@arfon 

Hello, the doi link for the paper [https://joss.theoj.org/papers/10.21105/joss.03852](https://joss.theoj.org/papers/10.21105/joss.03852) is showing a 404 not found message from github pages. The pdf download is available however.

**Comment by crvernon** [comment]
@jlarsen-usgs should be live now
